app/Main.hs

```diff
@@ -58,5 +58,36 @@ eval env (App t u) =
             Just value -> eval ((arg, value) : env') expr
         _                            -> Nothing
 
+list_env :: Env -> [Identifier]
+list_env []                = []
+list_env [(key, _)]        = [key]
+list_env ((key, _) : tail) = key : list_env tail
+
+delete :: (Eq a) => a -> [a] -> [a]
+delete _ []         = []
+delete x (y : tail) = if x == y
+                      then delete x tail
+                      else y : delete x tail
+
+into :: (Eq a) => a -> [a] -> Bool
+into _ [] = False
+into x (y : tail) = if x == y
+                    then True
+                    else into x tail
+
+intersect :: (Eq a) => [a] -> [a] -> [a]
+intersect [] _ = []
+intersect _ [] = []
+intersect (x : tail) y = if into x y
+                         then x : intersect tail y
+                         else intersect tail y
+
+
+unused :: Env -> Expr -> [Identifier]
```

Comment:
You can simply use `[Identifier]` instead of `Env` and pass it around in `unused`, then you wouldn't need `list_env`